src/java/org/apache/cassandra/db/RangeTombstoneList.java

```diff
@@ -58,12 +58,12 @@ public class RangeTombstoneList implements Iterable<RangeTombstone>, IMeasurable
     private ClusteringBound<?>[] starts;
     private ClusteringBound<?>[] ends;
     private long[] markedAts;
-    private int[] delTimes;
+    private long[] delTimes;
```

Comment:
This class seems to be able to internally store uint deletion times to save some memory, as we do with `DeletionTime`. Not sure if it is worth it, though.

Reply:
Yep the same can be said about every `long` TS in the codebase. But the underlying idea in the PR is to have a `long` codebase where we don't burden devs with dealing with units leaving that as a 'serialization' implementation detail. Also to not spread uints across the codebase like wildfire. `Cell` and `DeletionTime` being the exceptions.

At the end of the day we're trading a few bytes vs bit-wise operations. Being range tombstones... finger in the air uneducated guess... it might not be worth it but happy to be corrected :shrug: Like what's the usage here? Lot's of operations with many overlaps or many non-overlapping ranges with very little operations?...

Reply:
Not all long tombstones in the codebase are retained in the memtables. `DeletionTime` and the multiple implementations of `Cell` are, and so we have kept them as `int` to avoid increasing the memtable size. Aren't `RangeTombstoneList`s also retained in memtables, as part of partition's `DeletionInfo`? I don't know if the benchmark includes range tombstones.

Reply:
Just took a look at the last commit storing uint, which is very similar to what I was thinking. However, I think it misses the case where we add a range tombstone with a `InvalidDeletionTime`. 

On on hand, the valid-invalid state of an invalid deletion time is codified by the implemented class, so we can use a int for codifying the local deletion time. On the other hand, `RangeTombstoneList` seems to be an attempt at saving memory by aggregating multiple range tombstones, so we have an array of local deletion times. If this array is comprised by uints instead of longs, we miss a representation of the valid/invalid status that we used to represent with the class. 

I think this tiny test demonstrates it:
```java
// create an invalid deletion time
DeletionTime dt = DeletionTime.build(1, -1);
assertFalse(dt.validate());

// use the invalid deletion time for a range tombstone and aggregate it
RangeTombstoneList rtl = new RangeTombstoneList(null, 1);
rtl.add(new RangeTombstone(Slice.ALL, dt));

// undo the aggregation and see if the deletion time is still invalid
dt = rtl.iterator().next().deletionTime();
assertNotNull(dt);
assertFalse(dt.validate()); // fails
```
I guess we could try to overcome this problem by also storing an array of booleans on `RangeTombstoneList` storing what deletion times are invalid. Still, I'm not sure whether it's worth it and I wouldn't oppose to leaving `RangeTombstoneList`  with a `long` array. wdyt?

Reply:
I was waiting for CI to finish... But yeah if you add booleans then it doesn't make sense... :thinking: ...

Reply:
I don't have a better idea here. I'll revert. Can you please resolve all the rest you're happy with?

Reply:
I'm not sure adding the boolean array doesn't make sense since it would still save some memory and maybe conversions. But it might well not be a noticeable impact so I'm not against leaving it as it was.

As for resolving, reviewers cannot resolve conversations, just add a comment indicating agreement so the assignee can close them. I have left such comments on most conversations so you can close them.

However I'd keep this particular conversation open for a while, since there are other reviewers and they might find it interesting. With such a large PR it's easy to miss things so I think it's good to see what has caught the attention of other reviewers. I hope it's not too distracting to leave a few conversations open, specially if you use email notifications for tracking discussions.

Reply:
Sure let's leave this one open.

Iirc having a int and a boolean pair is going to be the same length (4+4) as having long (8). But we could have an int array of the invalid ldts indices in the first array. As this will be 99.9% of the time just a null, we shouldn't have invalids, so it seems like a possible solution.

I think I'll revert given we're entering the early optimizations world and at cost & complexity. The underlying idea here, imo, is that if this ever becomes a problem (or some other area of the code) we seem to be able to come up with reasonable alternatives.

Reply:
IIRC a boolean array takes one byte per element, plus the array overhead. That'd be 4+1 bytes per tombstone, lesser than the 8 bytes of using longs. Although taking advantage of the scarcity of invalid values to just have a set of invalid tombstone indexes sound good too.

I'm not yet convinced that trying to reduce the increase of the memtable size is early optimization. Do the benchmark on jira include range tombstones?

Reply:
Oh you mean just keep the bool and not the invalid ldt value too? Ok then yes it's not 8, correct.

Reducing memtable should be a thing yes. The benchmark had TTLs at row level and cell level. But I can try to come up with a more targeted test, yes.

Reply:
The magic of the weekend thinking: Managed to make RTL uint based with no extra arrays or anything :-) We already have a sentinel Uint for invalid ldts so just use that... wdyt? see latest commit.

Reply:
Great idea!